osquery/events/windows/windows_event_log.cpp

```diff
@@ -0,0 +1,175 @@
+                       nullptr,
+                       chan.c_str(),
+                       L"*",
+                       nullptr,
+                       nullptr,
+                       EVT_SUBSCRIBE_CALLBACK(winEventCallback),
+                       EvtSubscribeToFutureEvents);
+      if (hSubscription == nullptr) {
+        LOG(WARNING) << "Failed to subscribe to "
+                     << wstringToString(chan.c_str()) << ": ("
+                     << std::to_string(GetLastError()) << ")";
+      } else {
+        this->win_event_handles_.push_back(hSubscription);
+      }
+    }
+  }
+}
+
+void WindowsEventLogEventPublisher::restart() {
+  configure();
+}
+
+Status WindowsEventLogEventPublisher::run() {
+  if (this->win_event_handles_.empty()) {
+    configure();
+  }
+  pause();
+  return Status(0, "OK");
+}
+
+// TODO: Check this for mem/handle leaks
+void WindowsEventLogEventPublisher::stop() {
+  for (auto& e : this->win_event_handles_) {
+    if (e != nullptr) {
+      EvtClose(e);
+      e = nullptr;
+    }
+  }
+  win_event_handles_.clear();
+}
+
+void WindowsEventLogEventPublisher::tearDown() {
+  stop();
+}
+
+unsigned long __stdcall WindowsEventLogEventPublisher::winEventCallback(
+    EVT_SUBSCRIBE_NOTIFY_ACTION action, PVOID pContext, EVT_HANDLE hEvent) {
+  UNREFERENCED_PARAMETER(pContext);
+
+  unsigned long status = ERROR_SUCCESS;
+  pt::ptree propTree;
+  Status s;
+  switch (action) {
+  case EvtSubscribeActionError:
+    VLOG(1) << "Windows event callback failed with (" << hEvent << ")";
+    break;
+  case EvtSubscribeActionDeliver:
+    s = parseEvent(hEvent, propTree);
+    if (s.ok()) {
+      auto ec = createEventContext();
+      /// We leave the parsing of the properties up to the subscriber
+      ec->eventRecord = propTree;
+      ec->channel = stringToWstring(propTree.get("Event.System.Channel", ""));
+      EventFactory::fire<WindowsEventLogEventPublisher>(ec);
+    } else {
+      VLOG(1) << "Error rendering Windows event log (" << s.getCode() << ")";
+    }
+    break;
+
+  default:
+    VLOG(1) << "Received unknown action from Windows event log ("
+            << GetLastError() << ")";
+  }
+  return status;
+}
+
+Status WindowsEventLogEventPublisher::parseEvent(EVT_HANDLE evt,
+                                                 pt::ptree& propTree) {
+  unsigned long buffSize = 0;
+  unsigned long buffUsed = 0;
+  unsigned long propCount = 0;
+  LPWSTR xml = nullptr;
+  auto ret = EvtRender(
+      nullptr, evt, EvtRenderEventXml, buffSize, xml, &buffUsed, &propCount);
+  if (ret == 0) {
+    if (GetLastError() == ERROR_INSUFFICIENT_BUFFER) {
+      buffSize = buffUsed;
+      xml = static_cast<LPWSTR>(malloc(buffSize));
+      EvtRender(nullptr,
+                evt,
+                EvtRenderEventXml,
+                buffSize,
+                xml,
+                &buffUsed,
+                &propCount);
+    } else {
+      return Status(1, std::to_string(GetLastError()));
```

Comment:
I don't think `xml` will have been `malloc`ed if we return here? It only get's allocated if the `if` above is true. If we hit this spot `xml` is still a `nullptr` yeah?